### ISSUE TITLE
spdk nvmf initiator fails to connect while Linux initiator works well

### ISSUE THREAD

**Issue by sungjoonahn** [spdk/spdk#185]
Hi,

I have a nvmf setup with two CentOS 7 machines over Mellanox NIC cards. SPDK nvmf target works well with Linux initiator using nvme-cli for both physcal NVMe SSD and bdev malloc drive. But SPDK nvmf initiator fails while connecting to the same SPDK nvmf target. I tried perf and identify and they would exit with same error messages below. Can anyone tell me what to try?

SPDK version is v17.07 and Linux kernel is 4.12.8-1.el7.elrepo.x86_64 for both target and initiator machines.

Thanks,
Sungjoon

[perf]
/github/spdk/examples/nvme/perf/perf -q 1 -s 4096 -w read -M 100 -t 10 -r 'trtype:rdma adrfam:IPv4 traddr:10.0.0.1 trsvcid:4420 subnqn:nqn.2017-08.svcb-0011.spdk0'
Ignoring -M option... Please use -M option only when using rw or randrw.
Starting DPDK 17.05.0 initialization...
[ DPDK EAL parameters: perf -c 0x1 --no-pci --file-prefix=spdk_pid11079 ]
EAL: Detected 4 lcore(s)
Initializing NVMe Controllers
Attaching to NVMe over Fabrics controller at 10.0.0.1:4420: nqn.2017-08.svcb-0011.spdk0
nvme_rdma.c: 697:nvme_rdma_qpair_connect: *ERROR*: rdma_create_event_channel() failed
nvme_rdma.c:1270:nvme_rdma_ctrlr_construct: *ERROR*: failed to create admin qpair
nvme.c: 320:nvme_ctrlr_probe: *ERROR*: Failed to construct NVMe controller
...

[identify] 
/github/spdk/examples/nvme/identify/identify -r 'trtype:RDMA adrfam:IPv4 traddr:10.0.0.1 trsvcid:4420 subnqn:nqn.2017-08.svcb-0011.spdk0'
Starting DPDK 17.05.0 initialization...
[ DPDK EAL parameters: identify -c 0x1 -n 1 -m 64 --no-pci --file-prefix=spdk_pid13220 ]
EAL: Detected 4 lcore(s)
nvme_rdma.c: 697:nvme_rdma_qpair_connect: *ERROR*: rdma_create_event_channel() failed
nvme_rdma.c:1270:nvme_rdma_ctrlr_construct: *ERROR*: failed to create admin qpair
nvme.c: 320:nvme_ctrlr_probe: *ERROR*: Failed to construct NVMe controller

**Comment by wuwenjuncool** [comment]
Hi,
This error seem you don't have install the libmlx4 or libmlx5 ,please try : "dnf install libmlx4 libmlx5".
There is all  package you should install in Centos7 if you want to run perf with RDMA NIC:
sudo dnf install -y gcc gcc-c++ make CUnit-devel libaio-devel openssl-devel git astyle-devel python-pep8 lcov python clang-analyzer 
dnf install -y libibverbs-devel librdmacm-devel
sudo dnf install -y doxygen mscgen
sudo dnf install qperf 
sudo dnf install libmlx4 libmlx5

Thanks
Wenzhong

**Comment by sungjoonahn** [comment]
Hi Wenzhong,

The mellanox libraries, libmlx4 and libmlx5, had been installed. I also checked the other libraries you listed. I had all the libraries except astyle-devel. I think this library is irrelevant to this issue.

See messages from my setup about libmlx4 and libmlx5.

Thanks,
Sungjoon

[root@svcb-0011 spdk]# yum list | grep libmlx4
libmlx4.x86_64                          1.2.1-1.el7                    @base    
libmlx4.i686                            1.2.1-1.el7                    base     
libmlx4-static.i686                     1.2.1-1.el7                    base     
libmlx4-static.x86_64                   1.2.1-1.el7                    base     
[root@svcb-0011 spdk]# yum list | grep libmlx5
libmlx5.x86_64                          1.2.1-8.el7                    @base    
libmlx5.i686                            1.2.1-8.el7                    base     
libmlx5-static.i686                     1.2.1-8.el7                    base     
libmlx5-static.x86_64                   1.2.1-8.el7                    base     
[root@svcb-0011 spdk]# yum install libmlx4
Loaded plugins: fastestmirror, langpacks
base                                                                        | 3.6 kB  00:00:00     
elrepo                                                                      | 2.9 kB  00:00:00     
epel/x86_64/metalink                                                        |  12 kB  00:00:00     
epel                                                                        | 4.3 kB  00:00:00     
extras                                                                      | 3.4 kB  00:00:00     
updates                                                                     | 3.4 kB  00:00:00     
(1/3): epel/x86_64/group_gz                                                 | 170 kB  00:00:00     
(2/3): epel/x86_64/updateinfo                                               | 815 kB  00:00:01     
(3/3): epel/x86_64/primary_db                                               | 4.8 MB  00:00:02     
Loading mirror speeds from cached hostfile
 * base: mirror.sjc02.svwh.net
 * elrepo: ftp.osuosl.org
 * epel: mirrors.kernel.org
 * extras: mirror.sjc02.svwh.net
 * updates: mirror.sjc02.svwh.net
Package libmlx4-1.2.1-1.el7.x86_64 already installed and latest version
Nothing to do
[root@svcb-0011 spdk]# yum install libmlx5
Loaded plugins: fastestmirror, langpacks
Loading mirror speeds from cached hostfile
 * base: mirror.web-ster.com
 * elrepo: ftp.osuosl.org
 * epel: mirrors.kernel.org
 * extras: mirror.web-ster.com
 * updates: mirror.web-ster.com
Package libmlx5-1.2.1-8.el7.x86_64 already installed and latest version
Nothing to do_

**Comment by wuwenzhx** [comment]
Hi,Sungjoon

I also run perf with Kernel 4.12.8-1.el7.elrepo.x86_64 in my CentOS 7 machines,With the kernel 4.12.8-1.el7.elrepo.x86_64 I can't run perf with over Mellanox NIC cards too. But I can run perf  over Mellanox NIC cards with Kernel 4.10.1 and 3.10.1. I think you can try to change your kernel version.  

For me，I use kernel 4.10.1 everything is okay：
`[root@localhost spdk]# uname -r
4.10.1
[root@localhost spdk]# /root/spdk/examples/nvme/perf/perf -q 128 -s 4096 -w randread -t 15  -r 'trtype:RDMA adrfam:IPv4 traddr:192.168.9.11 trsvcid:4420 subnqn:nqn.2016-06.io.spdkcnode1'
Starting DPDK 17.05.0 initialization...
[ DPDK EAL parameters: perf -c 0x1 --no-pci --file-prefix=spdk_pid5590 ]
EAL: Detected 44 lcore(s)
EAL: No free hugepages reported in hugepages-1048576kB
Initializing NVMe Controllers
Attaching to NVMe over Fabrics controller at 192.168.9.11:4420: nqn.2016-06.io.spdkcnode1
Attached to NVMe over Fabrics controller at 192.168.9.11:4420: nqn.2016-06.io.spdkcnode1
Associating INTEL SSDPE2MD016T4  (PHFT648000111P6JGN  ) with lcore 0
Initialization complete. Launching workers.
Starting thread on core 0
========================================================
                                                                                            Latency(us)
Device Information                                     :       IOPS       MB/s    Average        min        max
INTEL SSDPE2MD016T4  (PHFT648000111P6JGN  ) from core 0:  418215.06    1633.65     305.99      38.43    4853.07
========================================================
Total                                                  :  418215.06    1633.65     305.99      38.43    4853.07

`

Thanks
Wenzhong

**Comment by sungjoonahn** [comment]
Hi Wenzhong,

Thanks to your help I found a working solution for my setup. Summary is below.

Thanks,
Sungjoon

===================
-Installed kernel 4.10.0 on both target and initiator. The kernel source was cloned and compiled from Torvalds's Linux github. The .config file was from CentOS 7 elrepo install version 4.13.x kernel.
-Compiled latest spdk for both target and initiator. The version I checked out was v17.07-342-ge2a3f1c.
-Make sure to load following kernel modules.
: spdk initiator - modprobe mlx5_ib mlx4_ib rdma_ucm
: spdk target - modprobe mlx5_ib rdma_ucm
Verified with perf, identify, and fio plugin. All seemed to work fine.
